# Regular Expressions

A Regular Expression (or Regex) is a pattern (or filter) that describes a set of strings that matches the pattern. In other words, a regex accepts a certain set of strings and rejects the rest.

### Functions in Regex
   - Search : Used to search a perticular pattern
   - Match : Used to find if RE matches at the start of the string
   - Finall : Used to find all the sub strings where RE is matched and returned as a list.
   - finditer :  Used to find all the sub strings where RE is matched and returned as an iterator
   - sub : It is used to find a perticular pattern and replace it with the desired pattern.

### Search Function

In [1]:
import re

In [2]:
text = "India, officially the Republic of India (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the Indian Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] China, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the Indian Ocean, India is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."
text

"India, officially the Republic of India (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the Indian Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] China, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the Indian Ocean, India is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."

In [6]:
re.search('Indonesia',text)

<re.Match object; span=(631, 640), match='Indonesia'>

span=(631, 640) indicates the word Indonesia starts from letter no. 631 and ends at letter no. 640

In [8]:
## Printing the Matched Pattern(If the name is available in the text, it print s the name else throws up an error)

name = re.search('Indonesia',text)
name.group()

'Indonesia'

In [23]:
def find_patterns(text,pattern):
    if re.search(pattern,text):
        return re.search(pattern,text)
    else:
        print('Pattern not found !!')

In [24]:
find_patterns(text,'America')

Pattern not found !!


In [25]:
find_patterns(text,'India')

<re.Match object; span=(0, 5), match='India'>

##### Quantifier

######  '?': Tells us if the specified pattern is present or not

In [29]:
def find_patterns(text,pattern):
    if re.search(pattern,text):
        return re.search(pattern,text)
    else:
        print('Pattern not found !!')

In [32]:
find_patterns("India","ac?")

<re.Match object; span=(4, 5), match='a'>

Show's there's only 1 match "a" and that is between position 4 and 5 in the word.

In [36]:
find_patterns("India","nd?")

<re.Match object; span=(1, 3), match='nd'>

In [40]:
find_patterns("INDIA","IA?")

<re.Match object; span=(0, 1), match='I'>

IT ONLY CHECKS TILL THE 1ST MATCH IS FOUND AND DOESNOT CARE IF THERE MAY BE MORE ACCURATE MATCHES

######  '*' Zero or More 

In [41]:
find_patterns("India","ac*")

<re.Match object; span=(4, 5), match='a'>

In [42]:
find_patterns("India","nd*")

<re.Match object; span=(1, 3), match='nd'>

In [49]:
find_patterns("INNNNDIIIA","DI*")

<re.Match object; span=(5, 9), match='DIII'>

######   '+' One or More

In [50]:
find_patterns("India","ac+")

Pattern not found !!


In [51]:
find_patterns("India","nd+")

<re.Match object; span=(1, 3), match='nd'>

In [54]:
find_patterns("INDIAAAA","IA+")

<re.Match object; span=(3, 8), match='IAAAA'>

###### '{}' Finding matches exact number of times

In [59]:
find_patterns("AApleAAAA","eA{4}")

<re.Match object; span=(4, 9), match='eAAAA'>

In [60]:
find_patterns("IAAAAppleIAA","IA{3}")

<re.Match object; span=(0, 4), match='IAAA'>

In [61]:
find_patterns("IAAAAPPLEIAA","IA{5}")

Pattern not found !!


###### {m,n} Finding in a range

In [62]:
find_patterns("IAAAAAAAAAAPPLEIAA","IA{3,7}")

<re.Match object; span=(0, 8), match='IAAAAAAA'>

In [64]:
find_patterns("IAAAAAAAAAAPPLEIAA","IA{4,}")

<re.Match object; span=(0, 11), match='IAAAAAAAAAA'>

#### Anchors

###### '^' Starts with

In [65]:
find_patterns("jack","^j")

<re.Match object; span=(0, 1), match='j'>

In [66]:
find_patterns("jack","^k")

Pattern not found !!


###### '$' Ends with

In [68]:
find_patterns("jack","k$")

<re.Match object; span=(3, 4), match='k'>

In [69]:
find_patterns("jack","z$")

Pattern not found !!


#### Wildcard '.' find any character

In [70]:
find_patterns("jack",".")

<re.Match object; span=(0, 1), match='j'>

#### Character Sets '[ ]' 
match anything present in the [ ] with the text.

In [74]:
find_patterns("jack","[ak]")

<re.Match object; span=(1, 2), match='a'>

In [81]:
find_patterns("jack","[a-k]") ## Searching in a range

<re.Match object; span=(0, 1), match='j'>

In [83]:
find_patterns("jaCk","[A-K]") ## Searching in a range

<re.Match object; span=(2, 3), match='C'>

###### Doesnot belong to '[^]'

In [85]:
find_patterns("jaCk","[^A-K]")

<re.Match object; span=(0, 1), match='j'>

In [86]:
find_patterns("jaCk","[^a-z]")

<re.Match object; span=(2, 3), match='C'>

In [88]:
find_patterns("jackz","[^a-j]")

<re.Match object; span=(3, 4), match='k'>

In [89]:
find_patterns("jaCk1","[^a-zA-K]")

<re.Match object; span=(4, 5), match='1'>

In [92]:
find_patterns("jaCk1","[^a-z0-9]")

<re.Match object; span=(2, 3), match='C'>

In [93]:
find_patterns("jaCk1!","[^a-z0-9A-Z]")

<re.Match object; span=(5, 6), match='!'>

#### Meta Sequeces

   - \s search for spaces
   - \s no spaces
   - \d == [0-9] searchec for digits only
   - \D == [^0-9] searches for everything other than digits
   - \w == [a-zA-Z0-9] Searches for alphabets and numbers only
   - \W ==[^a-zA-Z0-9] Searches for everything other than alphabets and numbers.

### Match Function : 
Used to find if RE matches at the start of the string

In [94]:
def match_patterns(text,pattern):
    if re.match(pattern,text):
        return re.search(pattern,text)
    else:
        print('Pattern not found !!')

In [97]:
match_patterns("jaCk1","j")

<re.Match object; span=(0, 1), match='j'>

In [98]:
match_patterns("jaCk1","1")

Pattern not found !!


### Sub Function: 
It is used to find a perticular pattern and replace it with the desired pattern. It needs 3 parameters:
   1. the text to be replaced
   2. replacement text
   3. Input text

In [101]:
input_text = "India, officially the Republic of India (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the Indian Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] China, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the Indian Ocean, India is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."
input_text

"India, officially the Republic of India (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the Indian Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] China, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the Indian Ocean, India is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."

In [102]:
re.sub("India","## My Country ##",input_text)

"## My Country ##, officially the Republic of ## My Country ## (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the ## My Country ##n Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] China, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the ## My Country ##n Ocean, ## My Country ## is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."

###### Using wild cards +sub

In [104]:
re.sub("Chi\w","***CHINA***", input_text)

"India, officially the Republic of India (Hindi: Bhārat Gaṇarājya),[26] is a country in South Asia. It is the seventh-largest country by area, the second-most populous country, and the most populous democracy in the world. Bounded by the Indian Ocean on the south, the Arabian Sea on the southwest, and the Bay of Bengal on the southeast, it shares land borders with Pakistan to the west;[f] ***CHINA***a, Nepal, and Bhutan to the north; and Bangladesh and Myanmar to the east. In the Indian Ocean, India is in the vicinity of Sri Lanka and the Maldives; its Andaman and Nicobar Islands share a maritime border with Thailand, Myanmar, and Indonesia. The nation's capital city is New Delhi."

### Findall Function :
Used to find all the sub strings where RE is matched and returned as a list. It returns only the first matching item.

In [112]:
text = "he is a monster 12/24/2022"
pattern = '\d{1,4}'
re.findall(pattern,text)

['12', '24', '2022']

In [119]:
### Exteacting from url's:
text = 'http://localhost:8888/notebooks-2022/10/31/Regular%20Expressions(regex)/2022/4/2.ipynb'
pattern = '/(\d{4})/(\d{1,2})/(\d{1,2})/'
re.findall(pattern,text)

[]

### Finditer Function: 
Used to find all the sub strings where RE is matched and returned as an iterator.
Used to find multmiple occurrences.

In [125]:
text = 'Today is sunday and on sunday I paly football'
pattern = 'sunday'

for match in re.finditer(pattern,text):
    print(match.span())

(9, 15)
(23, 29)


In [ ]:
## Semantic Processing :

Making machine understand what a sentence/paragrarph means or what is about.